ssl/quic/quic_dummy_handshake.c

```diff
@@ -0,0 +1,343 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/macros.h>
+#include <openssl/objects.h>
+#include "internal/quic_dummy_handshake.h"
+
+#define QUIC_DHS_MSG_TYPE_CH            0x01
+#define QUIC_DHS_MSG_TYPE_SH            0x02
+#define QUIC_DHS_MSG_TYPE_EE            0x08
+#define QUIC_DHS_MSG_TYPE_CERT          0x0B
+#define QUIC_DHS_MSG_TYPE_CERT_VERIFY   0x0F
+#define QUIC_DHS_MSG_TYPE_FINISHED      0x14
+
+#define QUIC_DHS_STATE_INITIAL              0
+#define QUIC_DHS_STATE_SENT_CH              1
+#define QUIC_DHS_STATE_RECEIVED_SH          2
+#define QUIC_DHS_STATE_RECEIVED_EE_HDR      8
+#define QUIC_DHS_STATE_RECEIVED_EE          3
+#define QUIC_DHS_STATE_RECEIVED_CERT        4
+#define QUIC_DHS_STATE_RECEIVED_CERT_VERIFY 5
+#define QUIC_DHS_STATE_RECEIVED_FINISHED    6
+#define QUIC_DHS_STATE_SENT_FINISHED        7
+
+#define QUIC_DHS_STATE_ERROR        0xFF
+
+struct quic_dhs_st {
+    QUIC_DHS_ARGS   args;
+    unsigned char   state;
+    unsigned char   *remote_transport_params;
+    size_t          remote_transport_params_len;
+    const unsigned char *local_transport_params;
+    size_t          local_transport_params_len;
+    unsigned char   rx_hdr[4];
+    size_t          rx_hdr_bytes_read;
+    size_t          rx_ee_bytes_read;
+};
+
+QUIC_DHS *ossl_quic_dhs_new(const QUIC_DHS_ARGS *args)
+{
+    QUIC_DHS *dhs;
+
+    if (args->crypto_send_cb == NULL
+        || args->crypto_recv_cb == NULL)
+        return NULL;
+
+    dhs = OPENSSL_zalloc(sizeof(*dhs));
+    if (dhs == NULL)
+        return NULL;
+
+    dhs->args   = *args;
+    dhs->state  = QUIC_DHS_STATE_INITIAL;
+    return dhs;
+}
+
+void ossl_quic_dhs_free(QUIC_DHS *dhs)
+{
+    if (dhs == NULL)
+        return;
+
+    OPENSSL_free(dhs->remote_transport_params);
+    OPENSSL_free(dhs);
+}
+
+static int dhs_send(QUIC_DHS *dhs, unsigned char type,
+                    const void *buf, size_t buf_len)
+{
+    size_t consumed = 0;
+    uint32_t len;
+    unsigned char hdr[4];
+
+    len = buf_len;
+    hdr[0] = type;
+    hdr[1] = (len >> 16) & 0xFF;
+    hdr[2] = (len >>  8) & 0xFF;
+    hdr[3] = (len      ) & 0xFF;
+
+    if (!dhs->args.crypto_send_cb(hdr, sizeof(hdr), &consumed,
+                                  dhs->args.crypto_send_cb_arg)
+        || consumed < sizeof(hdr)
+        || (buf_len > 0 && (!dhs->args.crypto_send_cb(buf, buf_len, &consumed,
+                                                      dhs->args.crypto_send_cb_arg)
+                            || consumed < buf_len)))
+        /*
+         * We do not handle a full buffer here properly but the DHS produces so
+         * little data this should not matter. By the time we want to fix this
+         * the real handshake layer will be ready.
+         */
+        return 0;
+
+    return 1;
+}
+
+static int dhs_recv_sof(QUIC_DHS *dhs, uint32_t *type, size_t *frame_len)
```

Comment:
This and other functions are missing documentation.

Reply:
They are static, we do not require documentation for static functions.

Reply:
https://github.com/openssl/technical-policies/blob/master/policies/documentation-policy.md#chapter-5-static-functions-and-globals-and-local-structures-and-macros

Reply:
Will be added in subsequent PR in any case.